File: mpi/integrate3.f90
program main
   use MPI
   implicit none
   integer :: size, rank, i, ierror
   double precision :: a, b, res, mya, myb, psum
   double precision, external :: integrate

   call MPI_Init(ierror)
   call MPI_Comm_size(MPI_COMM_WORLD, size, ierror)
   call MPI_Comm_rank(MPI_COMM_WORLD, rank, ierror)

   ! integration limits
   a = 0.0d0 ; b = 2.0d0 ; res = 0.0d0

   ! limits for "me"
   mya = a + rank*(b-a)/size
   myb = mya + (b-a)/size

   ! integrate f(x) over my own chunk - actual work
   psum = integrate(mya,myb)

   call MPI_Reduce(psum, res, 1, MPI_DOUBLE_PRECISION, MPI_SUM, 0, &
                   MPI_COMM_WORLD, ierror)

   if(rank.eq.0)then
      write(*,*)'Result =', res
   endif

   call MPI_Finalize(ierror)
end program main

! Return integral of cos(x) from x=a to x=b
function integrate(a, b)
   implicit none
   double precision :: a, b, integrate
   integrate = sin(b) - sin(a)
end function integrate

File: mpi/integrate3.c
#include <stdio.h>
#include <math.h>
#include <mpi.h>

// Return integral of cos(x) from x=a to x=b
double integrate(double a, double b)
{
   return sin(b) - sin(a);
}

// Compute integral(x=a to x=b) f(x)
int main(int argc, char** argv)
{
   int size, rank, i, ierr;
   double a, b, res, mya, myb, psum;

   MPI_Init(&argc, &argv);
   MPI_Comm_size(MPI_COMM_WORLD, &size);
   MPI_Comm_rank(MPI_COMM_WORLD, &rank);

   // integration limits
   a = 0.0; b = 2.0; res = 0.0;

   // limits for "me"
   mya = a + rank*(b-a)/size;
   myb = mya + (b-a)/size;
   // integrate f(x) over my own chunk - actual work
   psum = integrate(mya, myb);

   MPI_Reduce(&psum, &res, 1, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD);
   if(rank == 0) printf("Result: %20.14e\n",res);

   MPI_Finalize();
}
